Marginal improvement by pipelining loads

KungFuJesus opened the pull request:
The ld1{4 reg} variant saves us instructions
and only adds 3 cycles of latency to load 3
more neon/asimd registers worth of data.

Before:
```
 Tool: minigzip   Levels: 0-9       
 Runs: 100        Trim worst: 40        

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.040/0.083/0.102/0.014        0.099/0.143/0.162/0.017      211,973,953
 1     44.414%      2.875/2.919/2.945/0.020        0.930/1.004/1.040/0.025       94,139,819
 2     35.521%      4.949/5.072/5.108/0.030        0.977/1.037/1.068/0.023       75,289,481
 3     33.882%      6.610/6.683/6.715/0.026        0.913/0.997/1.039/0.033       71,815,574
 4     33.175%      7.811/7.881/7.915/0.025        0.907/0.968/0.996/0.021       70,317,604
 5     32.661%      8.361/8.439/8.471/0.024        0.864/0.949/0.984/0.028       69,227,371
 6     32.507%   10.167/10.259/10.291/0.028        0.887/0.951/0.979/0.021       68,902,008
 7     32.255%   14.870/14.949/14.982/0.026        0.840/0.931/0.967/0.030       68,366,928
 8     32.167%   21.945/21.996/22.032/0.023        0.853/0.936/0.959/0.023       68,180,702
 9     31.887%   22.258/22.306/22.338/0.021        0.872/0.951/0.982/0.022       67,586,442

 avg1  40.848%                       10.059                          0.887
 avg2  45.386%                       11.176                          0.985
 tot                               6035.267                        532.020      865,799,882
```

After:

```
 Tool: minigzip   Levels: 0-9       
 Runs: 100        Trim worst: 40        

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.020/0.081/0.108/0.019        0.091/0.143/0.168/0.018      211,973,953
 1     44.414%      2.782/2.879/2.914/0.027        0.940/1.000/1.028/0.019       94,139,819
 2     35.521%      4.987/5.037/5.069/0.021        0.958/1.032/1.067/0.025       75,289,481
 3     33.882%      6.522/6.631/6.664/0.032        0.906/0.992/1.027/0.032       71,815,574
 4     33.175%      7.690/7.835/7.866/0.031        0.897/0.957/0.993/0.026       70,317,604
 5     32.661%      8.313/8.386/8.420/0.027        0.866/0.948/0.976/0.022       69,227,371
 6     32.507%   10.144/10.217/10.253/0.027        0.865/0.936/0.973/0.025       68,902,008
 7     32.255%   14.853/14.904/14.931/0.020        0.873/0.933/0.964/0.022       68,366,928
 8     32.167%   21.899/21.957/21.992/0.023        0.877/0.936/0.962/0.019       68,180,702
 9     31.887%   22.193/22.278/22.321/0.029        0.887/0.952/0.984/0.026       67,586,442

 avg1  40.848%                       10.021                          0.883
 avg2  45.386%                       11.134                          0.981
 tot                               6012.416                        529.747      865,799,882
```

KungFuJesus commented:
Tested this on an odroid n2

KungFuJesus commented:
Ugh, yet another intrinsic not supported until way too late with GCC.  I guess we can do a compiler check to see if this is supported and fall back to single loads?

nmoinvaz commented:
Maybe just ifdef check for older GCC version?

KungFuJesus commented:
> Maybe just ifdef check for older GCC version?

Yeah I probably could but that might leave clang behind.  Given Apple's newest direction, we should probably generically detect this intrinsic pre-build.  Seems silly that we have to, but when gcc 9.3.x is too old I think it might be necessary.

mtl1979 commented:
Configure time check is what we have done before for most optional features... Benefit is that we don't have to do separate checks for gcc and clang as already mentioned. For some devices gcc 6.5 is still the latest usable version due to incompatibility issues in gcc 7.x series and later. For same reason Ubuntu 16.04 is latest usable version for some devices and 18.04 for others.

KungFuJesus commented:
I'll probably get to this before too long, but it'll probably give us better resolution if I wait until after the benchmark PR goes through, since I can independently assess the functions that are doing this 64 byte load.

KungFuJesus commented:
I plan to get back to this some time this weekend, but I rebased and ran the benchmarks.  It takes quite a few repetitions on my odroid-n2 to have a reasonable P value.  This seems to be the case in pretty much everything I optimized on this SBC, though, so that's nothing new.  There seems to be a consistent, ever so slight gain here:

```
Comparing slide_hash/neon (from ./benchmark_zlib) to slide_hash/neon (from ./benchmark_zlib_new)
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0219         -0.0215         20321         19876         20136         19704
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0275         -0.0296         20219         19663         20085         19490
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0199         -0.0182         20205         19803         20001         19637
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0570         -0.0544         20382         19220         20194         19095
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0051         -0.0060         19792         19692         19637         19519
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0509         -0.0491         20502         19458         20293         19296
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0275         -0.0232         20546         19981         20337         19865
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0072         -0.0070         19999         19856         19819         19680
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0459         -0.0467         20183         19257         20041         19105
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0401         -0.0365         20442         19622         20254         19515
[slide_hash/neon vs. slide_hash/neon]/1024                        +0.0085         +0.0100         20917         21095         20707         20914
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0167         -0.0173         20039         19705         19878         19533
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0245         -0.0246         20283         19786         20118         19622
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0029         -0.0011         20226         20167         20024         20002
[slide_hash/neon vs. slide_hash/neon]/1024                        -0.0340         -0.0331         20297         19606         20098         19433
[slide_hash/neon vs. slide_hash/neon]/1024_pvalue                  0.0002          0.0003      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/1024_mean                   -0.0249         -0.0239         20290         19786         20108         19627
[slide_hash/neon vs. slide_hash/neon]/1024_median                 -0.0285         -0.0281         20283         19705         20098         19533
[slide_hash/neon vs. slide_hash/neon]/1024_stddev                 +0.6795         +0.7483           262           440           247           433
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0212         -0.0210         20372         19940         20181         19756
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0197         -0.0194         20255         19856         20074         19685
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0534         -0.0483         20598         19498         20356         19372
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0247         -0.0231         20628         20119         20421         19949
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0105         -0.0093         20301         20088         20113         19927
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0325         -0.0334         20742         20067         20592         19904
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0446         -0.0417         20386         19477         20183         19342
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0507         -0.0497         20949         19886         20743         19712
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0250         -0.0263         20669         20153         20509         19969
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0364         -0.0370         20788         20032         20599         19836
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0102         -0.0072         20463         20254         20259         20114
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0077         -0.0086         20444         20286         20261         20087
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0100         -0.0116         20379         20176         20225         19991
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0218         -0.0214         20589         20140         20402         19966
[slide_hash/neon vs. slide_hash/neon]/2048                        -0.0236         -0.0202         20618         20131         20403         19991
[slide_hash/neon vs. slide_hash/neon]/2048_pvalue                  0.0000          0.0000      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/2048_mean                   -0.0262         -0.0253         20545         20007         20355         19840
[slide_hash/neon vs. slide_hash/neon]/2048_median                 -0.0243         -0.0211         20589         20088         20356         19927
[slide_hash/neon vs. slide_hash/neon]/2048_stddev                 +0.2410         +0.1907           196           243           195           232
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0346         -0.0307         20911         20187         20687         20052
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0336         -0.0353         21404         20685         21213         20463
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0342         -0.0351         20931         20214         20766         20037
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0637         -0.0604         21349         19988         21122         19845
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0374         -0.0384         20953         20170         20765         19968
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0388         -0.0395         21015         20199         20827         20005
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0404         -0.0371         20983         20136         20768         19997
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0530         -0.0538         21379         20246         21175         20036
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0495         -0.0493         21459         20398         21249         20201
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0410         -0.0426         20948         20090         20798         19913
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0515         -0.0485         20963         19882         20739         19734
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0379         -0.0377         21272         20466         21071         20276
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0300         -0.0297         20961         20332         20760         20143
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0559         -0.0555         21392         20195         21222         20044
[slide_hash/neon vs. slide_hash/neon]/4096                        -0.0478         -0.0488         21456         20430         21261         20224
[slide_hash/neon vs. slide_hash/neon]/4096_pvalue                  0.0000          0.0000      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/4096_mean                   -0.0433         -0.0429         21158         20241         20961         20062
[slide_hash/neon vs. slide_hash/neon]/4096_median                 -0.0388         -0.0379         21015         20199         20827         20037
[slide_hash/neon vs. slide_hash/neon]/4096_stddev                 -0.1207         -0.2011           227           199           225           180
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0269         -0.0264         22174         21576         21929         21351
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0346         -0.0327         22302         21530         22053         21333
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0380         -0.0374         22310         21462         22137         21308
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0117         -0.0114         22236         21976         22007         21756
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0516         -0.0490         22667         21498         22397         21299
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0631         -0.0599         22716         21283         22497         21149
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0111         -0.0123         22136         21890         21965         21694
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0376         -0.0375         22645         21794         22418         21577
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0470         -0.0468         22445         21390         22233         21193
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0182         -0.0177         22565         22155         22380         21983
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0106         -0.0089         21970         21736         21730         21536
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0622         -0.0615         22805         21387         22579         21190
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0216         -0.0202         22278         21798         22091         21644
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0342         -0.0324         22385         21619         22132         21415
[slide_hash/neon vs. slide_hash/neon]/8192                        -0.0453         -0.0445         22319         21309         22093         21110
[slide_hash/neon vs. slide_hash/neon]/8192_pvalue                  0.0000          0.0000      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/8192_mean                   -0.0344         -0.0334         22397         21627         22176         21436
[slide_hash/neon vs. slide_hash/neon]/8192_median                 -0.0333         -0.0353         22319         21576         22132         21351
[slide_hash/neon vs. slide_hash/neon]/8192_stddev                 +0.0790         +0.0746           239           257           237           254
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0242         -0.0205         25175         24566         24862         24353
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0292         -0.0264         24463         23748         24236         23596
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0370         -0.0363         24894         23973         24696         23799
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0280         -0.0260         25086         24383         24809         24165
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0238         -0.0190         24950         24356         24682         24214
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0384         -0.0401         25163         24198         24963         23963
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0018         +0.0005         24631         24588         24349         24361
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0430         -0.0416         25180         24097         24927         23891
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0513         -0.0479         24986         23703         24714         23530
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0049         -0.0068         24485         24365         24299         24134
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0093         -0.0060         25163         24929         24871         24722
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0342         -0.0305         25057         24201         24783         24027
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0087         -0.0085         24547         24334         24329         24121
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0208         -0.0194         24999         24478         24738         24258
[slide_hash/neon vs. slide_hash/neon]/16384                       -0.0019         +0.0009         24737         24690         24450         24472
[slide_hash/neon vs. slide_hash/neon]/16384_pvalue                 0.0001          0.0002      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/16384_mean                  -0.0238         -0.0219         24901         24307         24647         24107
[slide_hash/neon vs. slide_hash/neon]/16384_median                -0.0252         -0.0234         24986         24356         24714         24134
[slide_hash/neon vs. slide_hash/neon]/16384_stddev                +0.2832         +0.2967           262           336           247           320
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0447         -0.0434         30004         28663         29694         28403
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0280         -0.0265         29730         28897         29485         28704
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0395         -0.0372         30369         29169         30023         28906
[slide_hash/neon vs. slide_hash/neon]/32768                       +0.0074         +0.0108         30030         30251         29645         29966
[slide_hash/neon vs. slide_hash/neon]/32768                       +0.0068         +0.0078         29966         30170         29700         29932
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0413         -0.0402         30156         28911         29838         28638
[slide_hash/neon vs. slide_hash/neon]/32768                       +0.0110         +0.0094         29781         30109         29548         29826
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0403         -0.0388         30802         29562         30426         29244
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0327         -0.0302         30694         29689         30382         29466
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0053         -0.0067         29909         29752         29675         29476
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0126         -0.0119         30287         29904         29948         29590
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0083         -0.0093         29690         29444         29442         29168
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0317         -0.0283         29765         28821         29438         28606
[slide_hash/neon vs. slide_hash/neon]/32768                       -0.0154         -0.0156         30202         29737         29913         29448
[slide_hash/neon vs. slide_hash/neon]/32768                       +0.0067         +0.0066         29487         29685         29210         29402
[slide_hash/neon vs. slide_hash/neon]/32768_pvalue                 0.0048          0.0079      U Test, Repetitions: 15 vs 15
[slide_hash/neon vs. slide_hash/neon]/32768_mean                  -0.0180         -0.0170         30058         29518         29758         29252
[slide_hash/neon vs. slide_hash/neon]/32768_median                -0.0107         -0.0098         30004         29685         29694         29402
[slide_hash/neon vs. slide_hash/neon]/32768_stddev                +0.3977         +0.4720           369           515           340           501
OVERALL_GEOMEAN                                                   -0.0285         -0.0275             0             0             0             0
```

KungFuJesus commented:
I _think_ that should work?  We'll see what CI says.  The main issue is that I'm not 100% sure that the 4 register variant of ld1 is supported outside of aarch64, but I suspect that the compiler will still emulate the neon data structures and type definitions when dealing with 4xwide types.

KungFuJesus commented:
Hmm, actually looks like it should be supported throughout the whole neon family:
https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[Neon]&f:@navigationhierarchiesarchitectures=[v7,A64]&q=vld1_u32_

So eh, something dumb on the compilers' part, I guess.

mtl1979 commented on fallback_builtins.h:
```diff
@@ -41,4 +41,24 @@ static __forceinline unsigned long long __builtin_ctzll(uint64_t value) {
 #endif
 #endif
+
+#if defined(_M_ARM64) || defined(_M_ARM)
```
This is wrong... This applies to Visual C++ only, but older gcc versions didn't support these either... I know at least gcc 9.4.0 and newer do...

KungFuJesus replied:
I kind of figured as much with the _M prefix.  Maybe I can use a define we setup with the configure/cmake harness?

KungFuJesus replied:
Hmm, I'll just use the SLIDEHASH one since that's where it's used.  Conceivably the correct thing to do might be to find a common preprocessor define for all compilers or maybe have a bunch of ||'s with other places it ends up being used.

KungFuJesus commented:
Oh fun stuff - the types aren't defined, either.  Ugh...well, I guess I'll emulate them in lieu of a great way to do this even with assembly (the compiler needs cooperation because it requires participation from the register allocator).  I _suppose_ with the proper extended asm it could be done (using the right keywords and mnemonics that it mapped args directly to registers) but I don't want to do it for MSVC.

KungFuJesus commented:
Apparently we just bypass any sort of configure / cmake for the nmake CI?  It seems that MSVC supports these intrinsics (at least for the versions being used here), so maybe this isn't a huge deal.  Their arm support hasn't even existed _all_ that long, so maybe they've even always had them?

KungFuJesus commented:
Aaaand gcc 9 has the type definition for uint16x8x4 but not the load and store functions.  Ugh, compilers really can be a complicated mess.

mtl1979 commented on win32/Makefile.a64:
```diff
@@ -29,6 +29,7 @@ WFLAGS  = \
 	-D_CRT_NONSTDC_NO_DEPRECATE \
 	-DUNALIGNED_OK \
 	-DUNALIGNED64_OK \
+	-DARM_NEON_HASLD4 \
 	-D_ARM64_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE=1 \
 	-DARM_FEATURES \
```
Maybe should keep these in some kind of lexical order... Usually underscore comes before anything else though...

mtl1979 commented on win32/Makefile.arm:
```diff
@@ -28,6 +28,7 @@ WFLAGS  = \
 	-D_CRT_SECURE_NO_DEPRECATE \
 	-D_CRT_NONSTDC_NO_DEPRECATE \
 	-DUNALIGNED_OK \
+	-DARM_NEON_HASLD4 \
 	-D_ARM_WINAPI_PARTITION_DESKTOP_SDK_AVAILABLE=1 \
 	-DARM_FEATURES \
```
Maybe should keep these in some kind of lexical order... Usually underscore comes before anything else though...

mtl1979 commented on arch/arm/slide_hash_neon.c:
```diff
@@ -28,18 +30,21 @@ static inline void slide_hash_chain(Pos *table, uint32_t entries, uint16_t wsize
-        p += 8;
+        p0 = vld1q_u16_x4(table); 
+        p1 = vld1q_u16_x4(table+32); 
+        p0.val[0] = vqsubq_u16(p0.val[0], v);
+        p0.val[1] = vqsubq_u16(p0.val[1], v);
+        p0.val[2] = vqsubq_u16(p0.val[2], v);
+        p0.val[3] = vqsubq_u16(p0.val[3], v);
+        p1.val[0] = vqsubq_u16(p1.val[0], v);
+        p1.val[1] = vqsubq_u16(p1.val[1], v);
+        p1.val[2] = vqsubq_u16(p1.val[2], v);
+        p1.val[3] = vqsubq_u16(p1.val[3], v);
```
This pretty much repeats four identical lines for both variables... Might be worth making it inlined function or macro even though none of the supported compilers have own inlined x4 functions/macros for anything except loads and stores...

KungFuJesus commented:
The clang debug failure (on x86) is not a result of this commit but one of undefined behavior for the upper 3/4's of a zmm register when aliased from an XMM register. This should be fixed after rebasing against my pull request for avx512 improvements (PR: #1121).

KungFuJesus commented:
Just rebased.  It seems there are waaay fewer CI's being launched.  Did something change?

Hmm, actions tab on my forked repo still seems to be launching jobs and they are all completing as expected (except of course the pigz nothreads one, which we'd expect).

Dead2 commented:
CI looks good to me, maybe it got fixed after you pushed more commits.
Btw, if you rebase now, you'll get the fix for the pigz nothreads CI failure as well.

I only did a very cursory review of this today, and it looks a lot better now. I'd prefer it if you squashed the commits though, since there isn't really any benefit to keeping them separate.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -63,4 +63,50 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+#endif // _M_ARM64
+
+/* GCC 9 actually has this definition, but not the
+ * load and stores that use them. Go figure. */
```
Remove the go figure.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -63,4 +63,50 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+#  include <arm64_neon.h>
+#else
+#  include <arm_neon.h>
+#endif // _M_ARM64
```
Not sure this comment at #endif is necessary because the #ifdef is so nearby.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -63,4 +63,50 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+typedef struct uint16x8x4_t {
+    uint16x8_t val[4];
+} uint16x8x4_t;
+#endif // GCC < 9 test
```
Not sure this comment at #endif is necessary because the #ifdef is so nearby.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -63,4 +63,50 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+#endif // GCC < 9 test
+
+static inline uint16x8x4_t vld1q_u16_x4(uint16_t *a) {
+    uint16x8x4_t ret = (uint16x8x4_t){ 
```
Formatting/spacing for array might need improvement.

KungFuJesus committed:
Marginal improvement by pipelining loads on NEON

The ld1{4 reg} variant saves us instructions
and only adds 3 cycles of latency to load 3
more neon/asimd registers worth of data.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -16,7 +16,7 @@ static __forceinline unsigned long __builtin_ctz(uint32_t value) {
     if (x86_cpu_has_tzcnt)
         return _tzcnt_u32(value);
-#endif
+#endif // X86_FEATURES
```
Not sure this comment at #endif is necessary because the #ifdef is so nearby.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -31,16 +31,16 @@ static __forceinline unsigned long long __builtin_ctzll(uint64_t value) {
     if (x86_cpu_has_tzcnt)
         return _tzcnt_u64(value);
-#endif
+#endif // X86_FEATURES
```
Not sure this comment at #endif is necessary because the #ifdef is so nearby.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -7,7 +7,7 @@
 #ifdef X86_FEATURES
 #  include "arch/x86/x86.h"
-#endif
+#endif // X86_FEATURES
```
Not sure this comment at #endif is necessary because the #ifdef is so nearby.

nmoinvaz commented on fallback_builtins.h:
```diff
@@ -63,4 +63,50 @@ static inline __m512i _mm512_zextsi128_si512(__m128i a) {
+} while (0)
+
+/* Have to check for hard float ABI on GCC/clang, but not 
+ * on MSVC (guess it's assumed to be true? */
```
Comment could use some cleaning up.

mtl1979 replied:
Visual C++ doesn't support 32-bit ARMv8, and only FPU targets it supports for ARMv7 are VFPv3 and VFPv4... It does confusingly have NEON headers, but the compiler itself has no option to enable NEON support or target ARMv8. I filed a bug sometime ago to add support for 32-bit ARMv8 and as such proper command-line support for enabling or disabling NEON explicitly.

KungFuJesus replied:
If I remember correctly, ARM has had some issues consistently naming their features.  VFPv4 I think later became NEON, but it's instruction set was sort of written on the assumption that it had some optionally DSP offload component to it.  It would not surprise me in the slightest if all modern ARM CPUs reinterpreted VFPv4 instructions as NEON on the fly.  Or it may even be that the opcodes 100% overlap and they just renamed the mnemonics.

Oh yeah and then of course ARM of aarch64 decided that they didn't like the NEON name and decided to call it asimd, instead.

mtl1979 replied:
We do actually have some people from ARM contributing to this repository, but as far I have understood, as ARM instruction set is basically incremental set of features, for me it looks like VFPv3 has less features than VFPv4 (includes FMA+half-precision) and anything after than has more features. ARMv8 only had two optional features, ACLE and NEON.

Main difference between VFP and NEON (Advanced SIMD) is that VFP does execute the instructions per element sequentially but NEON does execute them in parallel, so is a lot faster.

KungFuJesus replied:
I think you're mostly right, v4 added some extra instructions but I think NEON basically inherited all of those instructions but was an actual SIMD ISA in execution (looks like wikipedia says VFPv4 operated sequentially).  I don't _think_ NEON is an optional component of ARMv8, but ACLE definitely is (which is why people were sort of annoyed that the pi4 didn't implement those instructions).  

https://developer.arm.com/documentation/den0024/a/AArch64-Floating-point-and-NEON

lol, of course, confusingly, "specialized" devices can be without it.  They also had a fun quirk where they somehow murdered performance on the TBL instruction (the aarch64/aarch32 equivalent of the shuffle instructions).  I'm not sure if every Cortex SIP suffers from this, but A72 definitely did.

mtl1979 replied:
32-bit only ARMv8 processors don't always have NEON... There is also another vector extension (Helium) used by some later ARMv8.1 processors that don't support NEON.

KungFuJesus replied:
Yes, but from what I understand helium was made for lower powered CPUs bearing the -M moniker.

mtl1979 replied:
True.... And that leaves us ARMv8-A and ARMv8-M when it comes to possible variations. With ARMv7 and earlier, there was many more variations, not forgetting VFPv3, VFPv4 and VFPv5 for the co-processor.

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -141,6 +141,34 @@ macro(check_neon_intrinsics)
+            endif()
+        endif()
+    endif()
+    # Check whether compiler supports NEON flag
```
Comment here needs to be updated.

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -141,6 +141,34 @@ macro(check_neon_intrinsics)
+        #  include <arm_neon.h>
+        #endif
+        int main(void) {
+            int stackVar[16];
```
Fix camelcase variable name.

nmoinvaz commented on configure:
```diff
@@ -1143,6 +1143,29 @@ EOF
+#  include <arm_neon.h>
+#endif
+int main(void) {
+    int stackVar[16];
```
Fix camelcase variable name.

nmoinvaz commented:
That's all I got. Nice work so far.

Dead2 commented:
Could you perhaps mention Neon in the first line of the commit message? That makes it easier to identify what the commit touches.

I want to do a benchmark of this before merging

KungFuJesus commented:
> Could you perhaps mention Neon in the first line of the commit message? That makes it easier to identify what the commit touches.
> 
> I want to do a benchmark of this before merging

Done.  I suggest doing the slide_hash/neon benchmark before and after the commit.  On my SBC at least, I had to do several benchmark repetitions for it to not fail the null test with a highish P value (there was a lot of variance).  It was a consistent performance win though, and it definitely moves the mean lower.  It's not a huge gain because the disparate loads were overlapping some on my A72's pipeline, but the latency wasn't entirely folded like it was with the 4 register variant that loads to a sequential range.

You also need a compiler that supports vld1q_u16_x4 or it will just fall back to using 4 disparate loads.  For me, this required gcc 10.  Resulting generated code basically looks like this:

```
   11dbc:       4cdf2448        ld1     {v8.8h-v11.8h}, [x2], #64
   11dc0:       4c40245c        ld1     {v28.8h-v31.8h}, [x2]
   11dc4:       6e602d14        uqsub   v20.8h, v8.8h, v0.8h
   11dc8:       6e602d35        uqsub   v21.8h, v9.8h, v0.8h
   11dcc:       6e602d56        uqsub   v22.8h, v10.8h, v0.8h
   11dd0:       6e602d77        uqsub   v23.8h, v11.8h, v0.8h
   11dd4:       6e602f98        uqsub   v24.8h, v28.8h, v0.8h
   11dd8:       6e602fb9        uqsub   v25.8h, v29.8h, v0.8h
   11ddc:       6e602fda        uqsub   v26.8h, v30.8h, v0.8h
   11de0:       6e602ffb        uqsub   v27.8h, v31.8h, v0.8h
   11de4:       4c002434        st1     {v20.8h-v23.8h}, [x1]
   11de8:       91020021        add     x1, x1, #0x80
   11dec:       4c002458        st1     {v24.8h-v27.8h}, [x2]
```

KungFuJesus commented:
@Dead2 Just curious, what hardware did you plan on benchmarking this on?

Dead2 commented:
### RPI 3, aarch64, GCC 10

## Baseline 88f331e6a64653b533721a645f8253b78e7f391a
```
 Tool: minigzip   Levels: 0-9
 Runs: 32         Trim worst: 18

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.008/0.015/0.020/0.003        0.016/0.034/0.044/0.008       15,737,543
 1     54.165%      0.535/0.548/0.555/0.007        0.188/0.201/0.208/0.006        8,523,602
 2     43.877%      0.904/0.921/0.936/0.010        0.184/0.199/0.210/0.008        6,904,649
 3     42.389%      1.202/1.225/1.241/0.011        0.166/0.194/0.205/0.011        6,670,476
 4     41.647%      1.358/1.372/1.380/0.007        0.168/0.187/0.197/0.009        6,553,763
 5     41.217%      1.474/1.498/1.517/0.016        0.181/0.190/0.199/0.007        6,485,981
 6     41.037%      1.740/1.757/1.765/0.008        0.172/0.187/0.195/0.008        6,457,780
 7     40.778%      2.405/2.417/2.427/0.007        0.167/0.176/0.187/0.007        6,416,924
 8     40.704%      2.938/2.949/2.955/0.005        0.168/0.181/0.194/0.007        6,405,244
 9     40.409%      3.302/3.328/3.341/0.011        0.178/0.193/0.200/0.007        6,358,951

 avg1  48.623%                        1.603                          0.174
 avg2  54.026%                        1.781                          0.194
 tot                                224.415                         24.390       76,514,913

   text    data     bss     dec     hex filename
 116566    1664      16  118246   1cde6 libz-ng.so.2
```

## PR #1074 9e45f15cd7d3ff7bd734ceb205ff7d514d4a5182
```
 Tool: minigzip   Levels: 0-9
 Runs: 32         Trim worst: 18

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.008/0.016/0.020/0.004        0.024/0.038/0.044/0.006       15,737,543
 1     54.165%      0.533/0.542/0.548/0.005        0.191/0.203/0.210/0.005        8,523,602
 2     43.877%      0.900/0.925/0.933/0.009        0.187/0.200/0.210/0.008        6,904,649
 3     42.389%      1.199/1.222/1.237/0.013        0.177/0.197/0.207/0.007        6,670,476
 4     41.647%      1.356/1.376/1.386/0.009        0.178/0.190/0.199/0.006        6,553,763
 5     41.217%      1.492/1.504/1.516/0.007        0.178/0.192/0.201/0.007        6,485,981
 6     41.037%      1.741/1.757/1.767/0.009        0.164/0.184/0.192/0.007        6,457,780
 7     40.778%      2.405/2.419/2.435/0.009        0.164/0.175/0.184/0.005        6,416,924
 8     40.704%      2.943/2.950/2.956/0.004        0.164/0.177/0.191/0.008        6,405,244
 9     40.409%      3.297/3.321/3.331/0.009        0.174/0.186/0.200/0.009        6,358,951

 avg1  48.623%                        1.603                          0.174
 avg2  54.026%                        1.781                          0.193
 tot                                224.467                         24.379       76,514,913

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.008/0.017/0.020/0.004        0.016/0.034/0.040/0.007       15,737,543
 1     54.165%      0.528/0.542/0.550/0.007        0.190/0.204/0.212/0.006        8,523,602
 2     43.877%      0.902/0.918/0.932/0.009        0.184/0.198/0.206/0.007        6,904,649
 3     42.389%      1.210/1.231/1.239/0.008        0.189/0.197/0.204/0.005        6,670,476
 4     41.647%      1.355/1.374/1.385/0.009        0.174/0.191/0.200/0.007        6,553,763
 5     41.217%      1.489/1.508/1.522/0.010        0.175/0.187/0.193/0.005        6,485,981
 6     41.037%      1.753/1.759/1.765/0.004        0.163/0.182/0.193/0.009        6,457,780
 7     40.778%      2.397/2.416/2.425/0.007        0.164/0.179/0.189/0.006        6,416,924
 8     40.704%      2.940/2.950/2.960/0.006        0.161/0.175/0.188/0.008        6,405,244
 9     40.409%      3.317/3.327/3.341/0.009        0.183/0.195/0.203/0.006        6,358,951

 avg1  48.623%                        1.604                          0.174
 avg2  54.026%                        1.783                          0.193
 tot                                224.600                         24.367       76,514,913

   text    data     bss     dec     hex filename
 116590    1664      16  118270   1cdfe libz-ng.so.2
 ```

Dead2 commented:
RPI 1B, arm6l, GCC 8.3
Fails to compile with this PR
```
[  3%] Building C object CMakeFiles/zlib.dir/compare256.c.o
/usr/lib/ccache/cc -DARM_ASM_HWCAP -DARM_AUXV_HAS_CRC32 -DARM_AUXV_HAS_NEON -DARM_FEATURES -DARM_NEON_ADLER32 -DARM_NEON_CHUNKSET -DARM_NEON_SLIDEHASH -DHAVE_BUILTIN_CTZ -DHAVE_BUILTIN_CTZLL -DHAVE_POSIX_MEMALIGN -DHAVE_VISIBILITY_HIDDEN -DHAVE_VISIBILITY_INTERNAL -DWITH_GZFILEOP -DZLIB_DLL -D_LARGEFILE64_SOURCE=1 -D__USE_LARGEFILE64 -I/root/github/zlib-ng  -Wall -Wpedantic -falign-functions=64 -O2 -DNDEBUG -fPIC   -fno-semantic-interposition -mfloat-abi=hard -Wall -Wextra -Wpedantic -Wno-implicit-fallthrough -std=c11 -o CMakeFiles/zlib.dir/compare256.c.o   -c /root/github/zlib-ng/compare256.c
In file included from /root/github/zlib-ng/compare256.c:7:
/root/github/zlib-ng/fallback_builtins.h:91:16: error: redefinition of ‘struct uint16x8x4_t’
 typedef struct uint16x8x4_t {
                ^~~~~~~~~~~~
In file included from /root/github/zlib-ng/fallback_builtins.h:84,
                 from /root/github/zlib-ng/compare256.c:7:
/usr/lib/gcc/arm-linux-gnueabihf/8/include/arm_neon.h:429:16: note: originally defined here
 typedef struct uint16x8x4_t
                ^~~~~~~~~~~~
In file included from /root/github/zlib-ng/compare256.c:7:
/root/github/zlib-ng/fallback_builtins.h:93:3: error: conflicting types for ‘uint16x8x4_t’
 } uint16x8x4_t;
   ^~~~~~~~~~~~
In file included from /root/github/zlib-ng/fallback_builtins.h:84,
                 from /root/github/zlib-ng/compare256.c:7:
/usr/lib/gcc/arm-linux-gnueabihf/8/include/arm_neon.h:432:3: note: previous declaration of ‘uint16x8x4_t’ was here
 } uint16x8x4_t;
   ^~~~~~~~~~~~
make[2]: *** [CMakeFiles/zlib.dir/build.make:92: CMakeFiles/zlib.dir/compare256.c.o] Error 1
```

```
-- Using CMake version 3.16.3
-- ZLIB_HEADER_VERSION: 1.2.11
-- ZLIBNG_HEADER_VERSION: 2.1.0.devel
-- The C compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/lib/ccache/cc
-- Check for working C compiler: /usr/lib/ccache/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Arch detected: 'armv6'
-- Basearch of 'armv6' has been detected as: 'arm'
-- Performing Test HAVE_FLOATABI_SOFTFP
-- Performing Test HAVE_FLOATABI_SOFTFP - Failed
-- Performing Test HAVE_FLOATABI_HARD
-- Performing Test HAVE_FLOATABI_HARD - Success
-- ARM floating point arch: -mfloat-abi=hard
-- Performing Test FNO_LTO_AVAILABLE
-- Performing Test FNO_LTO_AVAILABLE - Success
-- Looking for sys/sdt.h
-- Looking for sys/sdt.h - not found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - done
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for posix_memalign
-- Looking for posix_memalign - found
-- Performing Test HAVE_NO_INTERPOSITION
-- Performing Test HAVE_NO_INTERPOSITION - Success
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_HIDDEN
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_HIDDEN - Success
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_INTERNAL
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_INTERNAL - Success
-- Performing Test HAVE_BUILTIN_CTZ
-- Performing Test HAVE_BUILTIN_CTZ - Success
-- Performing Test HAVE_BUILTIN_CTZLL
-- Performing Test HAVE_BUILTIN_CTZLL - Success
-- Performing Test HAVE_PTRDIFF_T
-- Performing Test HAVE_PTRDIFF_T - Success
-- Performing Test ARM_AUXV_HAS_CRC32
-- Performing Test ARM_AUXV_HAS_CRC32 - Failed
-- Performing Test ARM_HWCAP_HAS_CRC32
-- Performing Test ARM_HWCAP_HAS_CRC32 - Success
-- Performing Test ARM_AUXV_HAS_NEON
-- Performing Test ARM_AUXV_HAS_NEON - Success
-- Performing Test MFPU_NEON_AVAILABLE
-- Performing Test MFPU_NEON_AVAILABLE - Success
-- Performing Test NEON_HAS_LD4
-- Performing Test NEON_HAS_LD4 - Failed
-- Architecture-specific source files: arch/arm/armfeature.c;arch/arm/adler32_neon.c;arch/arm/chunkset_neon.c;arch/arm/slide_hash_neon.c
-- The following features have been enabled:

 * CMAKE_BUILD_TYPE, Build type: Release (default)
 * NEON_ADLER32, Support NEON instructions in adler32, using "-mfpu=neon"
 * NEON_SLIDEHASH, Support NEON instructions in slide_hash, using "-mfpu=neon"
 * WITH_GZFILEOP, Compile with support for gzFile related functions
 * ZLIB_ENABLE_TESTS, Build test binaries
 * WITH_SANITIZER, Enable sanitizer support
 * WITH_OPTIM, Build with optimisation
 * WITH_NEW_STRATEGIES, Use new strategies
 * WITH_MAINTAINER_WARNINGS, Build with project maintainer warnings
 * WITH_NEON, Build with NEON intrinsics

-- The following features have been disabled:

 * ZLIB_SYMBOL_PREFIX, Publicly exported symbols DO NOT have a custom prefix
 * ZLIB_COMPAT, Compile with zlib compatible API
 * ZLIB_DUAL_LINK, Dual link tests against system zlib
 * WITH_FUZZERS, Build test/fuzz
 * WITH_BENCHMARKS, Build test/benchmarks
 * WITH_NATIVE_INSTRUCTIONS, Instruct the compiler to use the full instruction set on this host (gcc/clang -march=native)
 * WITH_CODE_COVERAGE, Enable code coverage reporting
 * WITH_INFLATE_STRICT, Build with strict inflate distance checking
 * WITH_INFLATE_ALLOW_INVALID_DIST, Build with zero fill for inflate invalid distances
 * WITH_UNALIGNED, Support unaligned reads on platforms that support it
 * WITH_UNALIGNED64, Support unaligned 64-bit reads on platforms that support it
 * WITH_ACLE, Build with ACLE
 * INSTALL_UTILS, Copy minigzip and minideflate during install

-- Configuring done
-- Generating done
```

KungFuJesus commented:
Hmm, interesting, so GCC 8 also defines that type.  I wonder how far back we need to go before it doesn't.

KungFuJesus commented:
@Dead2 So I found that typedef definition as far back as GCC4.  I think they were always there, but more for transposing loads.  This obviously does a straight load in memory order.  Go ahead and try to compile now.  

A note: you will not see this benefit at all on GCC 8, you need 10 or newer, I don't know why GCC took years to expose this instruction as an intrinsic, it's been there from the beginning.  For compilers that can't emit this instruction, our intrinsic simply emulates the behavior (the loads are put into the struct as though they were loaded with a single instruction).

So, if benchmarking, at least test it with 10 in addition to 8.  The inline assembly necessary to generate this intrinsic may not be very easily expressible to do it a better way (it requires a contiguous range of available registers and so it needs some cooperation from the register allocator).

KungFuJesus commented:
> ### RPI 3, aarch64, GCC 10
> ## Baseline [88f331e](https://github.com/zlib-ng/zlib-ng/commit/88f331e6a64653b533721a645f8253b78e7f391a)
> ```
>  Tool: minigzip   Levels: 0-9
>  Runs: 32         Trim worst: 18
> 
>  Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
>  0    100.008%      0.008/0.015/0.020/0.003        0.016/0.034/0.044/0.008       15,737,543
>  1     54.165%      0.535/0.548/0.555/0.007        0.188/0.201/0.208/0.006        8,523,602
>  2     43.877%      0.904/0.921/0.936/0.010        0.184/0.199/0.210/0.008        6,904,649
>  3     42.389%      1.202/1.225/1.241/0.011        0.166/0.194/0.205/0.011        6,670,476
>  4     41.647%      1.358/1.372/1.380/0.007        0.168/0.187/0.197/0.009        6,553,763
>  5     41.217%      1.474/1.498/1.517/0.016        0.181/0.190/0.199/0.007        6,485,981
>  6     41.037%      1.740/1.757/1.765/0.008        0.172/0.187/0.195/0.008        6,457,780
>  7     40.778%      2.405/2.417/2.427/0.007        0.167/0.176/0.187/0.007        6,416,924
>  8     40.704%      2.938/2.949/2.955/0.005        0.168/0.181/0.194/0.007        6,405,244
>  9     40.409%      3.302/3.328/3.341/0.011        0.178/0.193/0.200/0.007        6,358,951
> 
>  avg1  48.623%                        1.603                          0.174
>  avg2  54.026%                        1.781                          0.194
>  tot                                224.415                         24.390       76,514,913
> 
>    text    data     bss     dec     hex filename
>  116566    1664      16  118246   1cde6 libz-ng.so.2
> ```
> 
> ## PR #1074 [9e45f15](https://github.com/zlib-ng/zlib-ng/commit/9e45f15cd7d3ff7bd734ceb205ff7d514d4a5182)
> ```
>  Tool: minigzip   Levels: 0-9
>  Runs: 32         Trim worst: 18
> 
>  Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
>  0    100.008%      0.008/0.016/0.020/0.004        0.024/0.038/0.044/0.006       15,737,543
>  1     54.165%      0.533/0.542/0.548/0.005        0.191/0.203/0.210/0.005        8,523,602
>  2     43.877%      0.900/0.925/0.933/0.009        0.187/0.200/0.210/0.008        6,904,649
>  3     42.389%      1.199/1.222/1.237/0.013        0.177/0.197/0.207/0.007        6,670,476
>  4     41.647%      1.356/1.376/1.386/0.009        0.178/0.190/0.199/0.006        6,553,763
>  5     41.217%      1.492/1.504/1.516/0.007        0.178/0.192/0.201/0.007        6,485,981
>  6     41.037%      1.741/1.757/1.767/0.009        0.164/0.184/0.192/0.007        6,457,780
>  7     40.778%      2.405/2.419/2.435/0.009        0.164/0.175/0.184/0.005        6,416,924
>  8     40.704%      2.943/2.950/2.956/0.004        0.164/0.177/0.191/0.008        6,405,244
>  9     40.409%      3.297/3.321/3.331/0.009        0.174/0.186/0.200/0.009        6,358,951
> 
>  avg1  48.623%                        1.603                          0.174
>  avg2  54.026%                        1.781                          0.193
>  tot                                224.467                         24.379       76,514,913
> 
>  Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
>  0    100.008%      0.008/0.017/0.020/0.004        0.016/0.034/0.040/0.007       15,737,543
>  1     54.165%      0.528/0.542/0.550/0.007        0.190/0.204/0.212/0.006        8,523,602
>  2     43.877%      0.902/0.918/0.932/0.009        0.184/0.198/0.206/0.007        6,904,649
>  3     42.389%      1.210/1.231/1.239/0.008        0.189/0.197/0.204/0.005        6,670,476
>  4     41.647%      1.355/1.374/1.385/0.009        0.174/0.191/0.200/0.007        6,553,763
>  5     41.217%      1.489/1.508/1.522/0.010        0.175/0.187/0.193/0.005        6,485,981
>  6     41.037%      1.753/1.759/1.765/0.004        0.163/0.182/0.193/0.009        6,457,780
>  7     40.778%      2.397/2.416/2.425/0.007        0.164/0.179/0.189/0.006        6,416,924
>  8     40.704%      2.940/2.950/2.960/0.006        0.161/0.175/0.188/0.008        6,405,244
>  9     40.409%      3.317/3.327/3.341/0.009        0.183/0.195/0.203/0.006        6,358,951
> 
>  avg1  48.623%                        1.604                          0.174
>  avg2  54.026%                        1.783                          0.193
>  tot                                224.600                         24.367       76,514,913
> 
>    text    data     bss     dec     hex filename
>  116590    1664      16  118270   1cdfe libz-ng.so.2
> ```

Right, so the benefits are going to be hard to measure, especially given how small an impact it has over the overall process.  A better, more concentrated look would be to run the slide_hash benchmarks.

Try:
`benchmark_zlib --benchmark_filter=slide_hash/neon  --benchmark_repetitions=20`
Try that for both before and after the PR.  Google benchmark also has some python scripts that'll use scipy's prob & stats capabilities to give you a good idea of how much the distribution is moved and the probability of it being a NULL test, but it requires grabbing some python from their repos to get that.  It's what I'm running above with compare.py.

Dead2 commented:
RPI 1B, armv6l, GCC 8.3

## Baseline 88f331e6a64653b533721a645f8253b78e7f391a
```
 Tool: minigzip   Levels: 0-9
 Runs: 10         Trim worst: 6

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.180/0.195/0.220/0.017        0.210/0.215/0.230/0.010       15,737,543
 1     54.185%      3.480/3.507/3.520/0.019        0.830/0.835/0.840/0.006        8,526,745
 2     43.874%      4.990/5.110/5.160/0.081        0.750/0.787/0.820/0.029        6,904,168
 3     42.394%      6.240/6.290/6.320/0.038        0.770/0.782/0.790/0.010        6,671,197
 4     41.693%      7.200/7.240/7.300/0.045        0.770/0.782/0.790/0.010        6,560,888
 5     41.216%      7.680/7.768/7.850/0.071        0.760/0.772/0.790/0.013        6,485,936
 6     41.038%      8.890/9.062/9.180/0.123        0.750/0.770/0.790/0.023        6,457,827
 7     40.778%   11.000/11.072/11.160/0.066        0.730/0.755/0.770/0.019        6,416,941
 8     40.704%   13.400/13.620/13.840/0.232        0.690/0.718/0.750/0.028        6,405,249
 9     40.409%   12.660/12.785/13.020/0.167        0.760/0.795/0.820/0.030        6,358,951

 avg1  48.630%                        7.665                          0.721
 avg2  54.033%                        8.517                          0.801
 tot                                306.600                         28.850       76,525,445

   text    data     bss     dec     hex filename
 104778     856      16  105650   19cb2 libz-ng.so.2
```

## PR #1074 8e0af0a596eb72ea446072c61bbbcb3e84546162
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.170/0.177/0.190/0.010        0.200/0.225/0.240/0.017       15,737,543
 1     54.185%      3.450/3.513/3.560/0.052        0.790/0.815/0.850/0.026        8,526,745
 2     43.874%      4.970/5.055/5.130/0.073        0.790/0.802/0.830/0.019        6,904,168
 3     42.394%      6.310/6.375/6.440/0.054        0.780/0.788/0.790/0.005        6,671,197
 4     41.693%      7.090/7.200/7.320/0.112        0.730/0.762/0.780/0.022        6,560,888
 5     41.216%      7.700/7.763/7.830/0.059        0.710/0.745/0.760/0.024        6,485,936
 6     41.038%      8.930/9.045/9.130/0.087        0.770/0.785/0.800/0.013        6,457,827
 7     40.778%   11.010/11.107/11.190/0.083        0.700/0.722/0.750/0.022        6,416,941
 8     40.704%   13.190/13.518/13.700/0.228        0.700/0.745/0.760/0.030        6,405,249
 9     40.409%   12.600/12.770/12.970/0.179        0.770/0.785/0.800/0.017        6,358,951

 avg1  48.630%                        7.652                          0.718
 avg2  54.033%                        8.502                          0.797
 tot                                306.090                         28.700       76,525,445

   text    data     bss     dec     hex filename
 104650     856      16  105522   19c32 libz-ng.so.2

Dead2 commented:
Just as important as measuring the speedup is checking for regressions, such as yesterdays bug.
Things looks good now though.

KungFuJesus commented:
Heh, well, anything on the RPi1 was going to be noise or just alignment changes in code blocks because it doesn't even use this slide hash implementation.  But it's good we caught a compiler issue.
